File: Sources/CountriesFlagsObjC/source/Countries/UIImage+SolomonIslands.m
//
//  UIImage+SolomonIslands.m
//  CountriesFlags
//

#import "Countries/UIImage+SolomonIslands.h"

#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (SolomonIslands)

+ (UIImage *)flagSolomonIslandsWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:0.0f
                                         green:81.0f/255.0f
                                          blue:186.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:252.0f/255.0f
                                         green:209.0f/255.0f
                                          blue:22.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorGreen = [UIColor colorWithRed:33.0f/255.0f
                                         green:91.0f/255.0f
                                          blue:51.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorGreen setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorBlue setFill];
    UIBezierPath* path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointZero];
    [path addLineToPoint:CGPointMake(0.0f, size.height)];
    [path addLineToPoint:CGPointMake(size.width, 0.0f)];
    [path closePath];
    [path fill];

    [colorYellow setStroke];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width, 0.0f)];
    [path addLineToPoint:CGPointMake(0.0f, size.height)];
    [path closePath];
    path.lineWidth = size.width / 100.0f * 4.37f;
    [path stroke];

    [colorWhite setFill];
    CGFloat starRadius = size.height / 100.0f * 10.0f;
    // center star
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 100.0f * 17.5f, size.height / 100.0f * 30.0f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    // top left star
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 100.0f * 7.65f, size.height / 100.0f * 13.0f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    // bottom left star
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 100.0f * 7.65f, size.height / 100.0f * 47.0f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    // top right star
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 100.0f * 27.35f, size.height / 100.0f * 13.0f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    // top right star
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 100.0f * 27.35f, size.height / 100.0f * 47.0f)
                        radius:starRadius
                          type:FlagStarTypeFivePointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+SolomonIslands.swift
//
//  UIImage+SolomonIslands.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Solomon_Islands "wikipedia"
     
     Generates the flag of Solomon Islands.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Solomon Islands.
     */
    internal static func flagSolomonIslands(with size: CGSize) -> UIImage? {
        let colorBlue = UIColor(red: 0.0,
                                green: 81.0 / 255.0,
                                blue: 186.0 / 255.0,
                                alpha: 1.0)
        let colorYellow = UIColor(red: 252.0 / 255.0,
                                  green: 209.0 / 255.0,
                                  blue: 22.0 / 255.0,
                                  alpha: 1.0)
        let colorGreen = UIColor(red: 33.0 / 255.0,
                                 green: 91.0 / 255.0,
                                 blue: 51.0 / 255.0,
                                 alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorGreen.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorBlue.setFill()
        let path = UIBezierPath()
        path.move(to: .zero)
        path.addLine(to: CGPoint(x: 0.0, y: size.height))
        path.addLine(to: CGPoint(x: size.width, y: 0.0))
        path.close()
        path.fill()

        colorYellow.setStroke()
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width, y: 0.0))
        path.addLine(to: CGPoint(x: 0.0, y: size.height))
        path.close()
        path.lineWidth = size.width / 100.0 * 4.37
        path.stroke()

        colorWhite.setFill()
        let starRadius = size.height / 100.0 * 10.0
        // center star
        UIBezierPath.star(at: CGPoint(x: size.width / 100.0 * 17.5, y: size.height / 100.0 * 30.0),
                          radius: starRadius).fill()
        // top left star
        UIBezierPath.star(at: CGPoint(x: size.width / 100.0 * 7.65, y: size.height / 100.0 * 13.0),
                          radius: starRadius).fill()
        // bottom left star
        UIBezierPath.star(at: CGPoint(x: size.width / 100.0 * 7.65, y: size.height / 100.0 * 47.0),
                          radius: starRadius).fill()
        // top right star
        UIBezierPath.star(at: CGPoint(x: size.width / 100.0 * 27.35, y: size.height / 100.0 * 13.0),
                          radius: starRadius).fill()
        // top right star
        UIBezierPath.star(at: CGPoint(x: size.width / 100.0 * 27.35, y: size.height / 100.0 * 47.0),
                          radius: starRadius).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
